### ISSUE TITLE
[HOLD for payment 2024-09-11] ][$500] Lack of maximum height for the `description` field for room hides other options on the page

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#37357]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.44-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @JmillsExpensify 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1709046310280799

## Action Performed:
1. Create a room in the workspace with lengthy description 
2. Open the room and view the details of the room

## Expected Result:
All the option in the details page should be visible restricting the max height of the description field

## Actual Result:
Lack of a max height for the description field obscures other options

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

![CleanShot 2024-02-27 at 13 38 38@2x](https://github.com/Expensify/App/assets/38435837/1a945ac7-cb31-4899-a094-0f2bb4809882)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b55a4d576f4c3f8a</li>
        <li>Upwork Job ID: 1762602581669859328</li>
        <li>Last Price Increase: 2024-03-26</li>
<li>Automatic offers: </li>
<ul>
<li>shubham1206agra | Reviewer | 0</li>
<li>brandonhenry | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @greg-schroeder</details>

**Comment by Tony-MK** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Lack of maximum height for the description field for room hides other options on the page

### What is the root cause of that problem?

We are using the `autoGrowHeight` prop.

https://github.com/Tony-MK/Expensify-App/blob/bc3e5aa4791ed61221a679191e3b9533ed93c891/src/pages/RoomDescriptionPage.tsx#L91

### What changes do you think we should make in order to solve the problem?


Remove the `autoGrowHeight` prop from `InputWrapper` component.

**Comment by ShridharGoel** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
For lengthy descriptions, other options go out of the screen.

### What is the root cause of that problem?
There is no height limit or lines limit in the below code in MenuItem
```
{!!title && (shouldRenderAsHTML || (shouldParseTitle && !!html.length)) && (
                                                <View style={styles.renderHTMLTitle}>
                                                    <RenderHTML html={getProcessedTitle} />
                                                </View>
                                            )}
```

### What changes do you think we should make in order to solve the problem?
We can pass a prop from details page, which can be set as a max height limit in the above code to ensure that description height is limited.

**Comment by jeremy-croff** [comment]
It's not gonna be _that_ easy guys :P 
context https://github.com/Expensify/App/pull/34150#issuecomment-1915928141
Seems we want a solve here for truncating html

**Comment by jeremy-croff** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We want to limit the number of lines for html content 

### What is the root cause of that problem?
Our migration from https://github.com/Expensify/App/pull/34150#issuecomment-1915928141 is anticipating markdown support

### What changes do you think we should make in order to solve the problem?
I think as we start supporting more markdown/html we need more tooling to parse over this.
We can leverage a library such as https://github.com/arendjr/text-clipper or implement our own to clip html

### What alternative solutions did you explore? (Optional)
I also tried to solve it in a hackish way, setting a max height, overflowing, and adding ellipses with a pseudo element. It's ugly.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
There is no maximum line for the report description.

### What is the root cause of that problem?
Simply we don't limit the maximum line. The report description is rendered with a RenderHTML component and there is no straightforward way to set the number of lines props to the text.
https://github.com/Expensify/App/blob/98461f987e08e1ee977bd60f4974513b84b3c93d/src/components/MenuItem.tsx#L515-L519

### What changes do you think we should make in order to solve the problem?

One way is to wrap the description with a new custom tag that will have a custom renderer that can receive a number of lines attribute.

1. Create a new TextRenderer
```
function TextRenderer({TDefaultRenderer, textProps, ...props}) {
    return <TDefaultRenderer {...props} textProps={{...textProps, numberOfLines: props.tnode.attributes.line}} />
}
```
2. Register it to the custom renderer in [HTMLRenderers/index.ts](https://github.com/Expensify/App/blob/98461f987e08e1ee977bd60f4974513b84b3c93d/src/components/HTMLEngineProvider/HTMLRenderers/index.ts#L15-L30)
```
text: TextRenderer,
```
and in [`BaseHTMLEngineProvider`](https://github.com/Expensify/App/blob/98461f987e08e1ee977bd60f4974513b84b3c93d/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.tsx#L29-L73)
```
text: HTMLElementModel.fromCustomModel({tagName: 'text', contentModel: HTMLContentModel.textual}),
```
3. In [`MenuItem`](https://github.com/Expensify/App/blob/98461f987e08e1ee977bd60f4974513b84b3c93d/src/components/MenuItem.tsx#L364), wrap the HTML with `<text>` tag.
```
return processedTitle ? `<comment><text line="${numberOfLinesTitle}">${processedTitle}</text></comment>` : '';
```
4. In [`ReportDetailsPage`](https://github.com/Expensify/App/blob/98461f987e08e1ee977bd60f4974513b84b3c93d/src/pages/ReportDetailsPage.tsx#L245-L253), pass `numberOfLinesTitle` with the agreed value
```
numberOfLinesTitle={3}
```

_Why we can't create a custom renderer for `<comment>` tag? That's because [`comment` is a `block` model](https://github.com/Expensify/App/blob/98461f987e08e1ee977bd60f4974513b84b3c93d/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.tsx#L45-L49), not `textual` which won't receive `textProps`_

**Comment by dragnoir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Full description displayed.

### What is the root cause of that problem?

We are using `RenderHTML` to display the discription. The component does not support clipping text.

### What changes do you think we should make in order to solve the problem?

We can wrap the component with a `Text` and `numberOfLines={1}` prop.

```diff
+ <Text  numberOfLines={1}>
  <RenderHTML  html={getProcessedTitle}  />
+ </Text>
/>
```

POC:


### What alternative solutions did you explore? (Optional)

**Comment by sarahRosannaBusch** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When a room description is long, it pushes menu items off the screen so the user has to scroll down to see them. This happens on all platforms where the screen/window is too short to display everything in the details panel.

### What is the root cause of that problem?
The room description's height is variable because it is determined by the length of the user input.

### What changes do you think we should make in order to solve the problem?
Move the room description below the menu items in ReportDetailsPage.tsx. This ensures the menu items are always visible and the layout stays consistent. The bottom of long descriptions will still be off the screen, but it will be visually obvious that the user can scroll down to see the full text.

This solution is very simple to implement and doesn't require any adding any tools, functionality, or conditions, so it won't cause regressions. However, if there are design docs or user manuals that show this screen, they will need to be updated.

![Screenshot_proposal#37357](https://github.com/Expensify/App/assets/17365682/27f6859d-11fa-447d-8ef3-7332682e9359)

### What alternative solutions did you explore? (Optional)
- limit number of lines being displayed; this would unnecessarily hide data on screens that are large enough to display the full description
- make the description view height flexible to fit in the available space; this text area would still either need to be scrollable or collapsible for long descriptions, which looks messy and is problematic because it is already acting as a button to open the description editor
- make CONST.REPORT_DESCRIPTION.MAX_LENGTH smaller; this would be restrictive for users who want a long description, and wouldn't reliably fix the issue since users can increase their font sizes for accessiblity

**Comment by sarahRosannaBusch** [comment]
Contributor details
Your Expensify account email: annarose0113@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/sarahrosannabusch

**Comment by bernhardoj** [comment]
@ishpaul777 sure, here is the [test branch](https://github.com/bernhardoj/Expensify/tree/fix/37357-report-description-max-line)

**Comment by bernhardoj** [comment]
Oh, you're right. I only test it with plain text 🤦

**Comment by dragnoir** [comment]
@ishpaul777 will you consider removing the option to render the description as HTML and just render as Text, as an option?

**Comment by sarahRosannaBusch** [comment]
@ishpaul777 can you clarify what you're looking for here? If the requirement is to allow the user to input a long description that supports html/md, then we can't simply restrict the height of the description field because it is already a button. It would be very bad UX to have an "expand description" button within a button, which is why I proposed a simple layout change. Unless we don't need to show the full description on the details page and want to force the user to navigate to the Room Description editor to view the full text?

**Comment by jeremy-croff** [comment]
@ishpaul777 
My proposal suggested a solution to either implement or leverage existing tooling to parse over HTML and reduce the length by clipping it and appending ellipses. I'm not sure why this is not a solution. I additionally linked a library that already does this as an example of how this is done. 
I don't exactly want to write that code that does this or write a library proposal until you can clarify how you want to proceed in solving this ticket.

**Comment by jeremy-croff** [comment]
My RCA also identified this as a new feature request.

**Comment by greg-schroeder** [comment]
@ishpaul777 is still going through proposal review

**Comment by jeremy-croff** [comment]
@ishpaul777 If clipping the plain text is an idea you want to proceed with I can continue to write a POC of this implementation.  I think, in general, my RCA was accurate by identifying the app changes and the decisions made that led to this bug.

Key points in what my intention was when suggesting "new tooling" or a library.
1. I'm observing larger scale support for markdown in our roadmap
2. I think to correctly clip/truncate markdown in all cases of HTML it requires a robust implementation, hinting at a common utility or library, but we would foresee re-using this for more cases
3. It could be an extension of an existing common expensify package 

I think because of these factors I reserved some ambiguity as the exact util/architecture/package should require some internal review and I would proceed with some confirmation. But the linked library serves as a good pulse check on what the code would look like, and is about 600 lines to solve for this generically ( all elements ). I think an in-house solution should follow similiar logic.


The previous solution, from @brandonhenry, I think won't cut it because it's assuming its markdown is separate from "\n," but that's not true. And even then can only truncate a whole line at a time.

Thanks! :)

**Comment by jeremy-croff** [comment]
> > I think to correctly clip/truncate markdown in all cases of HTML it requires a robust implementation, hinting at a common utility or library, but we would foresee re-using this for more cases
> 
> Actually we do have some utilities https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js can we reuse any of utility from here ??

This is exactly the one I was referring to for
"3. It could be an extension of an existing common expensify package"

It seems we could reuse some of the regex for identifying the html blocks but still need an overal implementation for truncating.

**Comment by jeremy-croff** [comment]
I am reviewing https://github.com/Expensify/App/blob/main/contributingGuides/HOW_TO_CREATE_A_PLAN.md, is this what you want me to do?

**Comment by shawnborton** [comment]
Truncating the text makes sense to me - how else would we handle this? cc @Expensify/design

**Comment by dubielzyk-expensify** [comment]
Yep, exactly what Shawn said. Also want the truncating

**Comment by greg-schroeder** [comment]
Adding to #vip-vsb per roadmap doc:

>Anything related to a DM, room, group chat, thread, or task is in scope for here

**Comment by puneetlath** [comment]
For my understanding, does this solution require us to add a new dependency?

**Comment by puneetlath** [comment]
@brandonhenry how's it going?

**Comment by mallenexpensify** [comment]
@brandonhenry per [CONTRIBUTING.md,](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#:~:text=New%20contributors%20are%20limited%20to%20working%20on%20one%20job%20at%20a%20time%2C%20however%20experienced%20contributors%20may%20work%20on%20numerous%20jobs%20simultaneously.)  
> New contributors are limited to working on one job at a time, however experienced contributors may work on numerous jobs simultaneously.

You currently have [6 issues open](https://github.com/Expensify/App/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+assignee%3Abrandonhenry+) with non merged or closed.  Please do not submit proposals for new issues until **all** six PRs have been merged.

**Comment by greg-schroeder** [comment]
Work on PR continues

**Comment by greg-schroeder** [comment]
Linked PR in review

**Comment by greg-schroeder** [comment]
> Linked PR in review

Same story! Hopefully we can get this merged this week!

**Comment by greg-schroeder** [comment]
PR remains in review

**Comment by greg-schroeder** [comment]
Hey @brandonhenry - it seems @NikkiWines requested some changes on your PR, are you able to address those?

**Comment by mallenexpensify** [comment]
Damn, sorry to hear @brandonhenry , hope everyone's safe and that you get power back soon.

**Comment by greg-schroeder** [comment]
Thanks for your perseverance here @brandonhenry!

**Comment by greg-schroeder** [comment]
Thanks for your effort here!

**Comment by greg-schroeder** [comment]
Setting back to `Daily` here as there's active progress and last updates from @NikkiWines indicate that we're close here

**Comment by greg-schroeder** [comment]
Thanks @brandonhenry

**Comment by mallenexpensify** [comment]
Note from @shubham1206agra 
> There's another PR we need to merge first before completing the issue. But the expensify-common is 10 versions behind as of now.

@greg-schroeder please keep in mind, this might take a bit, with the delay being no fault of @brandonhenry

**Comment by puneetlath** [comment]
Sounds good, makes sense. We can update the price for the additional functionality.

**Comment by greg-schroeder** [comment]
PR in review

**Comment by greg-schroeder** [comment]
@brandonhenry This is a very long thread at this point - do you mind linking to the relevant PR that's waiting to be deployed for this issue to be resolved for me? Thanks!

**Comment by greg-schroeder** [comment]
Thanks @shubham1206agra! So hard to find sometimes 😅

**Comment by greg-schroeder** [comment]
PR is on staging, awaiting deploy to prod

**Comment by puneetlath** [comment]
@brandonhenry hmm I'm not sure, but actually let's hold off for a sec. I'm not sure that's the best solution. Let me think about it a bit without being rushed for the deploy.

**Comment by puneetlath** [comment]
@brandonhenry I'm going to led @nkdengineer raise the PR since they found the proper root cause [here](https://github.com/Expensify/App/issues/48050#issuecomment-2311711164). I don't think your solution will work in the scenario where it's longer than 200 characters, but doesn't contain any HTML.

**Comment by puneetlath** [comment]
We support images in room descriptions?

**Comment by puneetlath** [comment]
Ahh I see. I say let's live with it for a bit to see how we like it before adding more enhancements.

**Comment by greg-schroeder** [comment]
Hmm - was this not picked up by automation or something? @puneetlath do you know?

**Comment by mallenexpensify** [comment]
@greg-schroeder automation has been failing often lately (Rory is working on a fix!)
[The PR](https://github.com/Expensify/App/pull/38890) was deployed to production, automation failed to post it though, the way to check is to look at [the deploy checklist,](https://github.com/Expensify/App/issues/48035) if the checklist is closed, it's been deployed to production.  Labels, title and issue owner updated

**Comment by greg-schroeder** [comment]
Payment summary:

Contributor: @brandonhenry - $500 - Offer sent via Upwork
C+: @shubham1206agra - $500 - Offer sent via Upwork

**Comment by puneetlath** [comment]
@greg-schroeder I discussed with both Brandon and Shubham and re-checked the PRs. We had multiple scope increases, so I think we should update the payout to be $1250.

**Comment by brandonhenry** [comment]
@greg-schroeder @puneetlath thanks, let me know if I need to do anything here

**Comment by greg-schroeder** [comment]
Oh, word. All right - I didn't realize as the issue title still shows $500. I can update the offers.

**Comment by greg-schroeder** [comment]
Hmm - I updated it - did it not go through? I can try again

**Comment by greg-schroeder** [comment]
Tried again. Modified offer for @shubham1206agra - added bonus payment for @brandonhenry
